# Sprint 3 Tasca M3 T02
## Exercici 1
**Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.**

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/DelayedFlights.csv', encoding='utf-8')
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)

In [3]:
df.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
0           0  2008      1           3          4   2003.0        1955   2211.0        2225            WN        335  N712SW              128.0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0      8.0          0                N         0           NaN           NaN       NaN            NaN                NaN
1           1  2008      1           3          4    754.0         735   1002.0        1000            WN       3231  N772SW              128.0           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0     10.0          0                N         0           NaN           NaN       NaN            NaN                NaN
2           2  2008      1           3          4    628.0         620    804.0         750            WN        448  N428WN               96.0            90.0     76.0      14.0       8.0    IND  BWI       515     3.0     17.0          0                N         0           NaN           NaN       NaN            NaN                NaN
3           4  2008      1           3          4   1829.0        1755   1959.0        1925            WN       3920  N464WN               90.0            90.0     77.0      34.0      34.0    IND  BWI       515     3.0     10.0          0                N         0           2.0           0.0       0.0            0.0               32.0
4           5  2008      1           3          4   1940.0        1915   2121.0        2110            WN        378  N726SW              101.0           115.0     87.0      11.0      25.0    IND  JAX       688     4.0     10.0          0                N         0           NaN           NaN       NaN            NaN                NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

**Unnamed: 0**: Una columna con un índice numérico de los registros.

**Year**: El año de la fecha del vuelo.

**Month**: El mes de la fecha del vuelo.

**DayofMonth**: El día del mes de la fecha del vuelo.

**DayOfWeek**: El día de la semana de la fecha del vuelo, donde 1 es lunes y 7 es domingo.

**DepTime**: La hora de salida real del vuelo.

**CRSDepTime**: La hora de salida programada del vuelo.

**ArrTime**: La hora de llegada real del vuelo.

**CRSArrTime**: La hora de llegada programada del vuelo.

**UniqueCarrier**: El código de identificación único de la aerolínea.

**FlightNum**: El número de vuelo.

**TailNum**: El número de cola de la aeronave.

**ActualElapsedTime**: Duración real del vuelo en minutos.

**CRSElapsedTime**: Duración del vuelo programado en minutos.

**AirTime**: El tiempo de vuelo en el aire en minutos.

**ArrDelay**: El retraso en la llegada en minutos. Un vuelo se cuenta como "a tiempo" si operó menos de 15 minutos después de la hora programada que se muestra en CRSElapsedTime.

**DepDelay**: El retraso en la salida en minutos.

**Origin**: El aeropuerto de origen del vuelo.

**Dest**: El aeropuerto de destino del vuelo.

**Distance**: La distancia del vuelo en millas.

**TaxiIn**: El tiempo de rodaje en la llegada en minutos.

**TaxiOut**: El tiempo de rodaje en la salida en minutos.

**Cancelled**: Indica si el vuelo fue cancelado (1) o no (0).

**CancellationCode**: El código de cancelación si el vuelo fue cancelado.

**Diverted**: Indica si el vuelo fue desviado a otro aeropuerto (1) o no (0).

**CarrierDelay**: El tiempo de retraso debido a la aerolínea en minutos.

**WeatherDelay**: El tiempo de retraso debido al clima en minutos.

**NASDelay**: El tiempo de retraso debido al Sistema Nacional de Aviación en minutos.

**SecurityDelay**: El tiempo de retraso debido a problemas de seguridad en minutos.

**LateAircraftDelay**: El tiempo de retraso debido a problemas con el avión en minutos.

Inicialmente, considero todas las columnas relevantes. Probablemente no las mantendría todas en un solo dataframe, pero en principio encuentro que todas son interesantes para su uso, ya que podemos obtener datos importantes de cada una. La única excepción a primera vista sería la columna 'Unnamed: 0', ya que un índice no será útil para el análisis. Antes de continuar, verifiquemos la presencia de valores nulos para identificar posibles columnas repletas de registros corruptos o inexistentes.

In [5]:
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Tras analizar los resultados, realizaremos un breve estudio para encontrar una explicación a las últimas 5 columnas, ya que aproximadamente uno de cada tres registros contiene un valor nulo. El resumen del dataset ya nos avisa que retrasos inferiores a los 15 minutos no son considerados retrasos, si a eso le añadimos los vuelos cancelados y los desviados podría ser la explicación a la aparición de estos valores nulos. Comprobemos esta hipótesis.

In [6]:
df[df['ArrDelay'] < 15].shape[0]

680883

In [7]:
df.Cancelled.value_counts()

0    1936125
1        633
Name: Cancelled, dtype: int64

In [8]:
df.Diverted.value_counts()

0    1929004
1       7754
Name: Diverted, dtype: int64

Efectivamente la suma de vuelos cancelados, desviados y con demoras inferiores a 15 minutos son los 689270 registros sin valores. De manera que procedemos a cambir esos valores nulos por 0.

In [9]:
df[['CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']] = df[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].fillna(0)

In [10]:
df.isnull().sum()

Unnamed: 0              0
Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
DepTime                 0
CRSDepTime              0
ArrTime              7110
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum                 5
ActualElapsedTime    8387
CRSElapsedTime        198
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
TaxiIn               7110
TaxiOut               455
Cancelled               0
CancellationCode        0
Diverted                0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
dtype: int64

Tras llegar a este punto, decido realizar un filtrado del dataset, seleccionando únicamente los vuelos que no han sido cancelados ni desviados, ya que representan una minoría y merecen un análisis independiente. Al hacerlo, buscamos reducir cualquier interferencia que puedan causar en el estudio de los retrasos de los vuelos restantes.

In [11]:
print("Cancelled: {:.2f}%".format((df['Cancelled'] == 1).mean() * 100))
print("Diverted: {:.2f}%".format((df['Diverted'] == 1).mean() * 100))

Cancelled: 0.03%
Diverted: 0.40%


In [12]:
df_filtrado = df[(df['Cancelled'] == False) & (df['Diverted'] == False)]

Volvemos a comprobar la presencia de valores nulos dentro del  nuevo dataframe.

In [13]:
df_filtrado.isnull().sum()

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              3
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

Una vez aquí solo nos quedan 3 valores nulos en la columna 'NailNum''. La verdad es que pensé que podía corresponder a un vuelo regular de línea y poner el número de cola del avión más utilizado, pero no es el caso :(
A continuación, procederemos a eliminar las columnas: 'Unnamed: 0', 'Cancelled', 'CancellationCode' y 'Diverted', ya que no las necesitaremos.

In [14]:
df_filtrado = df_filtrado.drop(columns=['Unnamed: 0', 'Cancelled', 'CancellationCode', 'Diverted'])

Ahora es tiempo de mirar si existen duplicados y eliminarlos si los hubiera.


In [15]:
print( df_filtrado.duplicated().sum())

2


In [16]:
df_filtrado = df_filtrado.drop_duplicates()

Acabado el preprocesado de datos, a continuación reduciremos el dataset a 200000 registros de manera aleatoria.

In [17]:
df_reducido = df_filtrado.sample(n=200000, random_state=15)

df_reducido

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay
1273494  2008      8           9          6   1257.0        1250   1535.0        1535            WN       3976  N231WN              158.0           165.0    147.0       0.0       7.0    ISP  TPA      1034     4.0      7.0           0.0           0.0       0.0            0.0                0.0
563051   2008      3           9          7   1637.0        1445   1953.0        1823            CO       1182  N16646              136.0           158.0     88.0      90.0     112.0    ORD  EWR       719    14.0     34.0           0.0           0.0      90.0            0.0                0.0
910653   2008      6          23          1   1335.0        1325   1940.0        1935            WN       3410  N617SW              245.0           250.0    230.0       5.0      10.0    SJC  MDW      1838     6.0      9.0           0.0           0.0       0.0            0.0                0.0
1669533  2008     11           6          4   2003.0        1906    208.0          52            UA       1126  N374UA              245.0           226.0    218.0      76.0      57.0    DEN  BOS      1754     8.0     19.0          57.0           0.0      19.0            0.0                0.0
1275045  2008      8          10          7   1335.0        1325   1435.0        1435            WN       2483  N919WN               60.0            70.0     51.0       0.0      10.0    PHX  SNA       338     4.0      5.0           0.0           0.0       0.0            0.0                0.0
...       ...    ...         ...        ...      ...         ...      ...         ...           ...        ...     ...                ...             ...      ...       ...       ...    ...  ...       ...     ...      ...           ...           ...       ...            ...                ...
1745621  2008     12          16          2   1233.0        1215   1440.0        1425            WN       3008  N220WN              127.0           130.0    110.0      15.0      18.0    STL  HOU       687     2.0     15.0          15.0           0.0       0.0            0.0                0.0
1254283  2008      7          18          5   1431.0        1405   1827.0        1814            CO        466  N75861              236.0           249.0    214.0      13.0      26.0    SJU  EWR      1608    11.0     11.0           0.0           0.0       0.0            0.0                0.0
1748076  2008     12          18          4   1516.0        1410   1615.0        1515            WN       3677  N766SW               59.0            65.0     37.0      60.0      66.0    BUR  LAS       223     8.0     14.0           8.0           0.0       0.0            0.0               52.0
1508965  2008      9          11          4   1933.0        1925   2032.0        2020            AA       1633  N412AA               59.0            55.0     32.0      12.0       8.0    DFW  AUS       190     8.0     19.0           0.0           0.0       0.0            0.0                0.0
858025   2008      5          18          7   1542.0        1530   1924.0        1925            AA       1720  N290AA              162.0           175.0    144.0      -1.0      12.0    DFW  IAD      1172     6.0     12.0           0.0           0.0       0.0            0.0                0.0

[200000 rows x 26 columns]

# Exercici 2
**Fes un informe complet del dataset:**

- **Resumeix estadísticament el dataset i les columnes d’interès.**

In [18]:
# resumen estadístico de las variables numéricas

df_reducido.describe().T

count         mean          std     min      25%     50%     75%     max
Year               200000.0  2008.000000     0.000000  2008.0  2008.00  2008.0  2008.0  2008.0
Month              200000.0     6.098390     3.486221     1.0     3.00     6.0     9.0    12.0
DayofMonth         200000.0    15.748430     8.758178     1.0     8.00    16.0    23.0    31.0
DayOfWeek          200000.0     3.988000     1.992199     1.0     2.00     4.0     6.0     7.0
DepTime            200000.0  1519.517825   452.002660     1.0  1203.00  1547.0  1901.0  2400.0
CRSDepTime         200000.0  1470.014100   424.870930     1.0  1135.00  1510.0  1820.0  2359.0
ArrTime            200000.0  1609.598595   550.211926     1.0  1315.00  1716.0  2031.0  2400.0
CRSArrTime         200000.0  1636.359415   464.735300     1.0  1325.00  1709.0  2015.0  2359.0
FlightNum          200000.0  2186.075380  1939.911137     1.0   615.75  1550.0  3421.0  9740.0
ActualElapsedTime  200000.0   133.288965    72.129016    17.0    80.00   116.0   165.0   663.0
CRSElapsedTime     200000.0   134.216010    71.410350     2.0    81.00   116.0   165.0   660.0
AirTime            200000.0   108.283885    68.724665     3.0    58.00    90.0   137.0   642.0
ArrDelay           200000.0    42.202880    57.285710   -68.0     9.00    24.0    56.0  1525.0
DepDelay           200000.0    43.129925    53.854855     6.0    12.00    24.0    53.0  1395.0
Distance           200000.0   765.428245   574.300633    11.0   337.00   606.0   999.0  4962.0
TaxiIn             200000.0     6.805180     5.272059     0.0     4.00     6.0     8.0   163.0
TaxiOut            200000.0    18.199900    14.233788     0.0    10.00    14.0    21.0   331.0
CarrierDelay       200000.0    12.436215    36.501162     0.0     0.00     0.0    10.0  1382.0
WeatherDelay       200000.0     2.402040    17.492161     0.0     0.00     0.0     0.0  1148.0
NASDelay           200000.0     9.589760    28.024528     0.0     0.00     0.0     6.0  1337.0
SecurityDelay      200000.0     0.048435     1.210065     0.0     0.00     0.0     0.0   131.0
LateAircraftDelay  200000.0    16.474510    36.216688     0.0     0.00     0.0    18.0  1316.0

- **Fes una anàlisi estadístic del que consideris rellevant.**

Al realizar el estudio estadístico, lo primero que busqué fue la presencia de registros absurdos, como un día de la semana superior a 7, una distancia negativa o exageradamente grande, una hora menor a 0 o mayor a 2400, entre otros. Sin embargo, parece que no existen valores corruptos en los datos. A continuación, haré un breve resumen por cada bloque de información.

En las columnas correspondientes al año, mes, día del mes y día de la semana, encontramos los valores esperados. En cuanto al día de la semana, la media es aproximadamente 4 y la desviación estándar indica que los valores están bastante dispersos. Con respecto al día del mes, ocurre algo similar, con una media ligeramente superior a 15.5 y una distribución dispersa. Es curioso notar que en la columna del mes del año, la media es casi 6, aunque debería estar más cercana a 6.5. Esto podría sugerir que hay ligeramente más vuelos en la primera mitad del año que en la segunda. En cuanto al año vemos que todos los vuelos corresponden al 2008.

En relación a las columnas DepTime, CRSDepTime, ArrTime y CRSArrTime, simplemente podemos comentar que el estudio estadístico nos proporciona una idea aproximada. Sin embargo, la forma en que se tratan las horas en estas columnas muestra valores incoherentes, como el caso de CRSDepTime que la media es 1470. Sería interesante crear nuevas columnas que representaran las horas en formato decimal, lo cual tendrá un mayor valor estadístico. Vemos igualmente que las medias son las centrales del día teniendo en cuenta que a ciertas horas de la noche no se programan vuelos.

La columna FlightNum no tiene relevancia estadística en sí misma. Quizás podríamos calcular la moda o simplemente identificar los números de vuelo más repetidos.

En cuanto a ActualElapsedTime, CRSElapsedTime y AirTime, observamos que la duración promedio de los vuelos es de aproximadamente 2 horas, pero existe una gran variabilidad, desde 15 minutos hasta 11 horas. Lo que destaca es la similitud entre el tiempo estimado y el real, lo que indica que una vez que el avión tiene la orden de salida, los tiempos están meticulosamente medidos y planificados para minimizar el tiempo en el aire.

Los retrasos, que son la parte más interesante del dataframe, muestran medias similares entre los retrasos en las salidas y en las llegadas. Como mencionamos anteriormente, no es posible recortar significativamente los tiempos en el aire ni prolongarlos en exceso. Por lo tanto, un retraso en la salida probablemente conllevará un retraso en la llegada. Sin embargo, los valores de los retrasos en las llegadas son más variables, incluso encontramos valores negativos que indican que los vuelos han llegado antes de la hora programada. Y como es lógico, ningún vuelo despega antes de su hora.

La distancia es un valor muy variable, que va desde los 30 kilómetros hasta casi 5000 kilómetros en el vuelo más largo. Sin embargo, si observamos los percentiles, podemos ver que el 50% de los vuelos se encuentran en el rango de 300 a 1000 kilómetros.

En cuanto a los tiempos de rodaje, llama la atención a primera vista que los tiempos en las llegadas son tres veces más largos que los de salida. Esto se explica por la prioridad de los despegues en los aeropuertos, es decir, dejen salir antes de entrar :-) Añadir simplemente que el 75% de los vuelos tienen horquillas de tiempo de rodaje muy ajustadas, mientras que existe una parte correspondientes a retrasos que pueden superar las 6 horas.

Para finalizar, al analizar los diferentes motivos de los retrasos resulta sorprendente observar que las causas meteorológicas son mínimas, con un promedio de tan solo 2 minutos por vuelo. Por motivos de seguridad, los retrasos son prácticamente inexistentes. Sin embargo, las compañías aéreas y problemas con la aeronave presentan una media de retrasos de 15 minutos cada una.

- **Troba quantes dades faltants hi ha per columna.**

In [19]:
df_reducido.isnull().sum()

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

Nos damos cuenta de que la suerte estuvo de nuestro lado y los posibles valores nulos en la columna 'TailNum' quedaron fuera. 
En caso de que hubiera quedado alguno, habríamos procedido a eliminarlo directamente o, en su defecto, reemplazarlo por un '0' para no perder los datos.

- **Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).**

In [20]:
df_reducido['AvgSpeed'] = df_reducido['Distance'] / (df_reducido['AirTime']/60)
df_reducido['AvgSpeed'].describe()

count    200000.000000
mean        397.039959
std          78.096918
min          21.290323
25%         352.000000
50%         404.000000
75%         448.405797
max        8080.000000
Name: AvgSpeed, dtype: float64

In [21]:
df_reducido[df_reducido['AvgSpeed'] > 1000]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay     AvgSpeed
236966   2008      2          12          2   1700.0        1440   1818.0        1629            OH       5517  N818CA               78.0           109.0     14.0     109.0     140.0    CMH  JFK       483    42.0     22.0           0.0           0.0     109.0            0.0                0.0  2070.000000
1483229  2008      9          15          1   1422.0        1415   1545.0        1539            EV       4609  N680BR               83.0            84.0      3.0       6.0       7.0    CAE  CVG       404    66.0     14.0           0.0           0.0       0.0            0.0                0.0  8080.000000
890906   2008      6           9          1   1301.0        1250   1823.0        1835            WN       2968  N664WN              202.0           225.0     76.0     -12.0      11.0    LAX  STL      1593   119.0      7.0           0.0           0.0       0.0            0.0                0.0  1257.631579
1243146  2008      7           1          2   1702.0        1630   1916.0        1753            B6         55  N665JB              134.0            83.0     16.0      83.0      32.0    BTV  JFK       267    10.0    108.0           0.0           0.0      51.0            0.0               32.0  1001.250000
235530   2008      2           4          1   1950.0        1940   2034.0        2050            OH       5486  N964CA               44.0            70.0      7.0     -16.0      10.0    CVG  DTW       229    22.0     15.0           0.0           0.0       0.0            0.0                0.0  1962.857143
939241   2008      6           7          6   2125.0        1925   2215.0        2130            OH       5515  N554CA               50.0           125.0     30.0      45.0     120.0    CVG  LGA       585     9.0     11.0          45.0           0.0       0.0            0.0                0.0  1170.000000
1308503  2008      8          29          5    820.0         755   1010.0        1035            YV       2668  N914FJ               50.0           100.0     16.0     -25.0      25.0    MEM  CLT       512    10.0     24.0           0.0           0.0       0.0            0.0                0.0  1920.000000
1315995  2008      8           5          2   1652.0        1459   1709.0        1557            OO       6041  N936SW               17.0            58.0      7.0      72.0     113.0    DEN  GJT       212     4.0      6.0           0.0          72.0       0.0            0.0                0.0  1817.142857

A partir de la velocidad media, se puede deducir que algunos valores de tiempo en aire no son correctos, ya que resulta absurdo que un avión haya estado en el aire solo 3 minutos y haya recorrido 404 millas. En cuanto a los valores bajos, esto podría atribuirse a vuelos de corta distancia que estaban volando cerca del destino, esperando permiso para el aterrizaje.

In [22]:
df_reducido['Delayed'] = df_reducido['ArrDelay'] >= 15
print("Delayed: {:.2f}%".format((df_reducido['Delayed'].mean() * 100)))

Delayed: 64.56%


Observamos que aproximadamente un tercio de los vuelos llega a destino puntual, considerando que se define como puntual un retraso de no más de 15 minutos, mientras que el resto acumula algún tipo de retraso.

Para facilitar el análisis y la visualización de la evolución diaria del número de vuelos a lo largo del año, vamos a crear dos nuevas columnas en el dataset. La primera columna se llamará 'Date' y contendrá la fecha en formato (yyyy/mm/dd). La segunda columna se llamará 'DayofYear' y representará el número del día en el año actual. Estas nuevas columnas nos permitirán graficar cómodamente la tendencia diaria de los vuelos a lo largo del año.

In [23]:
# parece ridículo, pero no he sido capaz que me lo aceptase si no cambiaba el nombre de la columna :(
df.rename(columns={'DayofMonth': 'Day'}, inplace=True)
df['Date'] = pd.to_datetime(df[['Year', 'Month', 'Day']])

df['DayofYear'] = df['Date'].dt.dayofyear
df.DayofYear.value_counts() 

2      10857
357    10693
358    10621
336    10367
356     9994
       ...  
264     1908
244     1849
278     1806
333     1736
332     1451
Name: DayofYear, Length: 366, dtype: int64

- **Fes una taula de les aerolínies amb més endarreriments acumulats.**

In [24]:
print(df_reducido.groupby('UniqueCarrier')['Delayed'].sum().reset_index().sort_values('Delayed', ascending=False).to_string(index=False))

UniqueCarrier  Delayed
           WN    21103
           AA    13721
           MQ    10193
           UA     9757
           OO     9233
           DL     7529
           XE     7374
           US     6259
           CO     6058
           EV     5903
           NW     5775
           YV     5210
           FL     4868
           OH     3983
           B6     3878
           9E     3681
           AS     2454
           F9     1667
           HA      441
           AQ       36


In [25]:
print(df_reducido.groupby('UniqueCarrier').agg(UniqueCarrier=('UniqueCarrier', 'first'), DelayedPercentage=('Delayed', 'mean')).sort_values('DelayedPercentage', ascending=False).to_string(index=False))

UniqueCarrier  DelayedPercentage
           OH           0.756362
           YV           0.750612
           B6           0.697733
           AA           0.696285
           NW           0.695448
           EV           0.693818
           XE           0.692524
           MQ           0.691848
           9E           0.685858
           OO           0.678099
           UA           0.672572
           FL           0.661144
           DL           0.634181
           AS           0.617514
           US           0.608615
           CO           0.588384
           F9           0.568554
           HA           0.564661
           AQ           0.537313
           WN           0.537286


Tenía que limpiar el nombre de SouthWest Airlines ('WN') puesto que acumula mas retrasos en terminos absolutos.Sin embargo si miramos los porcentajes respecto al total de vuelos, está entre las 3 aerolineas que acumulan menos retrasos.

- **Quins són els vols més llargs? I els més endarrerits?**

En esta sección, tomaremos en cuenta la duración de los vuelos para identificar cuáles son los más largos.

In [26]:
print(df_reducido.groupby('FlightNum')['CRSElapsedTime'].max().reset_index().sort_values('CRSElapsedTime', ascending=False).head(5).to_string(index=False))

 FlightNum  CRSElapsedTime
        15           660.0
      1273           604.0
       851           601.0
      1561           580.0
        14           575.0


In [27]:
print(df_reducido.groupby('FlightNum')['Delayed'].sum().reset_index().sort_values('Delayed', ascending=False).head(5).to_string(index=False))

 FlightNum  Delayed
        50      128
        16      115
       321      103
       378       95
       357       92


- **Busca les rutes més llargues i les que acumulen més retards.**

Sin embargo aquí hemos consideraremos la distancia de los vuelos. En la primera parte, hemos agregado una condición para evitar la repetición de la misma ruta en ambas direcciones, puesto que la distancia es la misma.

In [28]:
condicion = df_reducido['Origin'] < df_reducido['Dest']
print(df_reducido[condicion].groupby(['Origin', 'Dest'])['Distance'].max().reset_index().sort_values('Distance', ascending=False).head(5).to_string(index=False))

Origin Dest  Distance
   EWR  HNL      4962
   ATL  HNL      4502
   HNL  ORD      4243
   KOA  ORD      4213
   HNL  MSP      3972


In [29]:
print(df_reducido.groupby(['Origin', 'Dest'])['Distance'].sum().reset_index().sort_values('Distance', ascending=False).head(5).to_string(index=False))

Origin Dest  Distance
   JFK  LAX    650925
   SFO  JFK    630984
   ORD  LAX    537460
   LAX  JFK    534600
   JFK  SFO    496512


- **Aporta allò que consideris rellevant.**

Llegados a este punto, sería interesante analizar qué aeropuertos presentan un mayor porcentaje de minutos de retraso acumulados en relación con el número de vuelos operados.

In [30]:
tabla_origen = df.groupby('Origin')['DepDelay'].mean().reset_index()
print("Tabla con aeropuertos de origen con mas restrasos acumulados")
print(tabla_origen.sort_values(tabla_origen.columns[1], ascending=False).head(10).to_string(index=False))

Tabla con aeropuertos de origen con mas restrasos acumulados
Origin   DepDelay
   CMX 116.147059
   PLN  93.761905
   SPI  83.848739
   ALO  82.225806
   MQT  79.556650
   ACY  79.310345
   MOT  78.661654
   HHH  76.530055
   EGE  74.128920
   BGM  73.155340


In [31]:
print("Tabla con aeropuertos de origen con menos restrasos acumulados")
print(tabla_origen.sort_values(tabla_origen.columns[1], ascending=True).head(10).to_string(index=False))

Tabla con aeropuertos de origen con menos restrasos acumulados
Origin  DepDelay
   TUP 11.000000
   YKM 22.472727
   INL 23.000000
   GST 25.075000
   WYS 25.100000
   PUB 27.000000
   IPL 28.517007
   OXR 29.268966
   CLD 29.721271
   ADK 30.509434


In [32]:
tabla_destino = df.groupby('Dest')['ArrDelay'].mean().reset_index()
print("Tabla con aeropuertos de destino con mas restrasos acumulados")
print(tabla_destino.sort_values(tabla_destino.columns[1], ascending=False).head(10).to_string(index=False))

Tabla con aeropuertos de destino con mas restrasos acumulados
Dest  ArrDelay
 MQT 77.939252
 SPI 72.127877
 INL 66.777778
 TEX 66.266667
 ALO 65.944444
 MCN 65.674033
 CMX 63.530612
 ORD 60.028153
 EWR 59.877946
 ACK 59.167742


In [33]:
print("Tabla con aeropuertos de destino con menos restrasos acumulados")
print(tabla_destino.sort_values(tabla_destino.columns[1], ascending=True).head(10).to_string(index=False))

Tabla con aeropuertos de destino con menos restrasos acumulados
Dest  ArrDelay
 TUP  6.000000
 PIR 15.000000
 BJI 20.666667
 IYK 21.112903
 BLI 22.846154
 GST 23.326087
 WYS 23.741935
 HTS 23.848485
 IPL 24.828729
 YKM 26.744681


Ahora ya sabemos que aeropuertos evitar en el futuro y cuales son mas fiables. Para terminar haré una pequeña matriz de correlación para intentar de sacarle un poco mas de jugo a este dataframe.

In [34]:
df_reducido[['DayOfWeek', 'Distance', 'AirTime', 'AvgSpeed', 'Delayed']].corr()

DayOfWeek  Distance   AirTime  AvgSpeed   Delayed
DayOfWeek   1.000000  0.006649  0.003788  0.011121 -0.001871
Distance    0.006649  1.000000  0.980424  0.653927 -0.055876
AirTime     0.003788  0.980424  1.000000  0.546409 -0.017394
AvgSpeed    0.011121  0.653927  0.546409  1.000000 -0.153895
Delayed    -0.001871 -0.055876 -0.017394 -0.153895  1.000000

Como era de esperar, la distancia y el tiempo en el aire están estrechamente relacionados. Sin embargo, la velocidad media de los vuelos no muestra una gran correlación con la duración del vuelo, lo que indica que los aviones siguen estándares de velocidad y que la distancia parece tener un peso mayor. Esto probablemente se debe a que, para vuelos más largos, se utilizan aviones más grandes con mayor potencia, lo que les permite volar un poco más rápido.

Si queremos buscar las razones de los retrasos, tendremos que analizar más detalladamente. La tabla muestra que prácticamente no existe correlación con la distancia, el tiempo en el aire ni siquiera con la velocidad media, aunque esta última es ligeramente superior, lo que indica que hay un pequeño margen en el que, si el avión puede volar más rápido, puede evitar retrasos. Es importante destacar que los valores negativos indican una relación inversa, es decir, cuanto mayor sea la velocidad media, menor será el retraso.

En cuanto al día de la semana, obviamente no muestra correlación con ninguna de las variables, ni siquiera con los retrasos, a pesar de la creencia de que los fines de semana acumulan más retrasos.

## Exercici 3
**Exporta el dataset net i amb les noves columnes a Excel.**

In [35]:
df_reducido.to_excel('Vuelos_2008.xlsx')